File: android/framework_old/src/main/java/com/dreampany/nearby/core/Connection.java
package com.dreampany.nearby.core;

import android.content.Context;

import androidx.annotation.NonNull;

import com.dreampany.nearby.misc.Runner;
import com.dreampany.nearby.misc.SmartQueue;
import com.dreampany.nearby.util.NearbyUtil;
import com.google.android.gms.nearby.Nearby;
import com.google.android.gms.nearby.connection.AdvertisingOptions;
import com.google.android.gms.nearby.connection.ConnectionInfo;
import com.google.android.gms.nearby.connection.ConnectionLifecycleCallback;
import com.google.android.gms.nearby.connection.ConnectionResolution;
import com.google.android.gms.nearby.connection.ConnectionsClient;
import com.google.android.gms.nearby.connection.ConnectionsStatusCodes;
import com.google.android.gms.nearby.connection.DiscoveredEndpointInfo;
import com.google.android.gms.nearby.connection.DiscoveryOptions;
import com.google.android.gms.nearby.connection.EndpointDiscoveryCallback;
import com.google.android.gms.nearby.connection.Payload;
import com.google.android.gms.nearby.connection.PayloadCallback;
import com.google.android.gms.nearby.connection.PayloadTransferUpdate;
import com.google.android.gms.nearby.connection.Strategy;
import com.google.common.base.Strings;
import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import com.google.common.collect.Maps;

import java.util.Map;
import java.util.concurrent.Executor;

import timber.log.Timber;

/**
 * Created by Hawladar Roman on 6/4/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
class Connection extends ConnectionLifecycleCallback {

    private final Object guard = new Object();

    private enum State {
        FOUND, LOST,
        REQUESTING, REQUEST_SUCCESS, REQUEST_FAILED,
        INITIATED, ACCEPTED, REJECTED, DISCONNECTED
    }

    interface Callback {
        void onConnection(long peerId, boolean connected);

        void onPayload(long peerId, Payload payload);

        void onPayloadStatus(long peerId, PayloadTransferUpdate status);
    }

    private Context context;
    private Executor executor;
    private long serviceId;
    private long peerId;
    private Callback callback;
    private ConnectionsClient client;
    private AdvertisingOptions advertisingOptions;
    private DiscoveryOptions discoveryOptions;

    private volatile boolean advertising = false;
    private volatile boolean discovering = false;

    private BiMap<Long, String> cache;
    private BiMap<Long, String> endpoints; // peerId to endpointId
    private Map<String, State> states;     // endpointId to State
    private Map<String, Boolean> directs;  // endpointId to Directs (incoming = true or outgoing = false)
    private SmartQueue<String> pendingEndpoints;
    private Map<String, Integer> requestTries;
    private final int MAX_TRY = 5;

    private volatile RequestThread requestThread;

    Connection(Context context, Executor executor, long serviceId, long peerId, Strategy strategy, Callback callback) {
        this.context = context.getApplicationContext();
        this.executor = executor;
        this.serviceId = serviceId;
        this.peerId = peerId;
        this.callback = callback;
        this.client = Nearby.getConnectionsClient(this.context);
        this.advertisingOptions = new AdvertisingOptions.Builder().setStrategy(strategy).build();
        this.discoveryOptions = new DiscoveryOptions.Builder().setStrategy(strategy).build();

        cache = HashBiMap.create();
        endpoints = HashBiMap.create();
        states = Maps.newConcurrentMap();
        directs = Maps.newConcurrentMap();
        pendingEndpoints = new SmartQueue<>();
        requestTries = Maps.newConcurrentMap();
    }

    @Override
    public void onConnectionInitiated(@NonNull String endpointId, @NonNull ConnectionInfo info) {
        long peerId = toLong(info.getEndpointName());
        endpoints.put(peerId, endpointId);
        states.put(endpointId, State.INITIATED);
        directs.put(endpointId, info.isIncomingConnection());
        client.acceptConnection(endpointId, payloadCallback);
        Timber.v("First Connection initiated for EndpointId(%s) Peer(%d) incoming %s", endpointId, peerId, info.isIncomingConnection());
    }

    @Override
    public void onConnectionResult(@NonNull String endpointId, @NonNull ConnectionResolution result) {
        boolean accepted = result.getStatus().getStatusCode() == ConnectionsStatusCodes.STATUS_OK;
        //TODO STATUS_CONNECTION_REJECTED
        Timber.v("First Connection Result endpoint: %s accepted %s", endpointId, accepted);
        states.put(endpointId, accepted ? State.ACCEPTED : State.REJECTED);

        if (accepted) {
            pendingEndpoints.remove(endpointId);
            long peerId = getPeerId(endpointId);
            if (callback != null) {
                callback.onConnection(peerId, true);
            }
        } else {
            pendingEndpoints.insertLastUniquely(endpointId);
            startRequestThread();
        }
    }

    @Override
    public void onDisconnected(@NonNull String endpointId) {
        Timber.v("First Disconnected endpoint: %s", endpointId);
        states.put(endpointId, State.DISCONNECTED);
        pendingEndpoints.insertLastUniquely(endpointId);
        long peerId = getPeerId(endpointId);
        executor.execute(() -> {
            if (callback != null)
                callback.onConnection(peerId, false);
        });
        startRequestThread();
    }

    void start() {
        synchronized (guard) {
            startAdvertising();
            startDiscovery();
        }
    }

    void stop() {
        synchronized (guard) {
            stopRequestThread();
            stopAdvertising();
            stopDiscovery();
        }
    }

    private void startAdvertising() {
        synchronized (guard) {
            if (advertising) {
                return;
            }
            Timber.v("Advertising fired for Peer (%d) - ServiceId (%d)", peerId, serviceId);
            client.startAdvertising(
                    toString(peerId),
                    toString(serviceId),
                    this,
                    advertisingOptions)
                    .addOnSuccessListener(ignored -> {
                        advertising = true;
                        Timber.v("Success Advertising of Peer (%d) - ServiceId (%d)", peerId, serviceId);
                    })
                    .addOnFailureListener(error -> {
                        advertising = false;
                        Timber.e("Error in Advertising of Peer (%d) - ServiceId (%d) - %s", peerId, serviceId, error.getMessage());
                    });
        }
    }

    private void startDiscovery() {
        synchronized (guard) {
            if (discovering) {
                return;
            }
            Timber.v("Discovering fired for Peer (%d) - ServiceId (%d)", peerId, serviceId);
            client.startDiscovery(
                    toString(serviceId),
                    discoveryCallback,
                    discoveryOptions)
                    .addOnSuccessListener(ignored -> {
                        discovering = true;
                        Timber.v("Success Discovering Peer (%d) - ServiceId (%d)", peerId, serviceId);
                    })
                    .addOnFailureListener(error -> {
                        discovering = false;
                        Timber.e("Error Discovering Peer (%d) - ServiceId (%d) - %s", peerId, serviceId, error.getMessage());
                    });
        }
    }

    private void stopAdvertising() {
        synchronized (guard) {
            advertising = false;
            client.stopAdvertising();
        }
    }

    private void stopDiscovery() {
        synchronized (guard) {
            discovering = false;
            client.stopDiscovery();
        }
    }

    private void requestConnection(final String endpointId) {
        Timber.v("Requesting Connection: %s state[%s]", endpointId, states.get(endpointId));
        client.requestConnection(
                toString(peerId),
                endpointId,
                this)
                .addOnSuccessListener(ignored -> {
                    Timber.v("Request Connection succeed for (%s)", endpointId);
                    states.put(endpointId, State.REQUEST_SUCCESS);
                    pendingEndpoints.remove(endpointId);
                    requestTries.remove(endpointId);
                })
                .addOnFailureListener(error -> {
                    Timber.e("Request Connection error (%s) - %s", endpointId, error.getMessage());
                    Timber.e(error);
                    states.put(endpointId, State.REQUEST_FAILED);
                    pendingEndpoints.insertLastUniquely(endpointId);
                    startRequestThread();
                });
    }

    /* Connection life cycle callback */
    private final ConnectionLifecycleCallback connectionCallback = new ConnectionLifecycleCallback() {
        @Override
        public void onConnectionInitiated(@NonNull String endpointId, @NonNull ConnectionInfo info) {
            Timber.v("Second Connection initiated endpointId(%s) %s", endpointId, info.isIncomingConnection());

        }

        @Override
        public void onConnectionResult(@NonNull String endpointId, @NonNull ConnectionResolution result) {
            boolean accepted = result.getStatus().getStatusCode() == ConnectionsStatusCodes.STATUS_OK;
            Timber.v("Second Connection Result endpoint: %s accepted %s", endpointId, accepted);
        }

        @Override
        public void onDisconnected(@NonNull String endpointId) {
            Timber.v("Second Disconnected endpoint: %s", endpointId);
        }
    };

    /* Discovery callback for getting advertised devices */
    private final EndpointDiscoveryCallback discoveryCallback = new EndpointDiscoveryCallback() {

        @Override
        public void onEndpointFound(@NonNull String endpointId, @NonNull DiscoveredEndpointInfo info) {
            long serviceId = toLong(info.getServiceId());
            if (Connection.this.serviceId != serviceId) {
                Timber.e("Unknown ServiceId (%d) from EndpointId (%s)", serviceId, info.getEndpointName());
                return;
            }

            long peerId = toLong(info.getEndpointName());
            Timber.v("Found EndpointId (%s) - PeerId (%d)", endpointId, peerId);

            //priority works: remove old endpoints if exists
            if (endpoints.containsKey(peerId)) {
                String oldEndpointId = endpoints.remove(peerId);
                states.remove(oldEndpointId);
                directs.remove(oldEndpointId);
                pendingEndpoints.remove(oldEndpointId);
            }

            endpoints.put(peerId, endpointId);
            states.put(endpointId, State.FOUND);
            pendingEndpoints.insertLastUniquely(endpointId);
            requestTries.put(endpointId, 0);
            startRequestThread();
        }

        @Override
        public void onEndpointLost(@NonNull String endpointId) {
            long peerId = getPeerId(endpointId);
            Timber.v("Endpoint lost (%s) - PeerId (%d)", endpointId, peerId);
            endpoints.remove(peerId);
            states.put(endpointId, State.LOST);
            directs.remove(endpointId);
            pendingEndpoints.remove(endpointId);
            requestTries.remove(endpointId);
        }
    };

    //Payload callback
    private final PayloadCallback payloadCallback = new PayloadCallback() {
        @Override
        public void onPayloadReceived(@NonNull String endpointId, @NonNull Payload payload) {
            long peerId = endpoints.inverse().get(endpointId);
            Timber.v("Payload Received from PeerId (%d)", peerId);
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    if (callback != null) {
                        callback.onPayload(peerId, payload);
                    }
                }
            });
        }

        @Override
        public void onPayloadTransferUpdate(@NonNull String endpointId, @NonNull PayloadTransferUpdate update) {
            long peerId = endpoints.inverse().get(endpointId);
            Timber.v("Payload Transfer Update from PeerId (%d)", peerId);
            executor.execute(() -> {
                if (callback != null) {
                    callback.onPayloadStatus(peerId, update);
                }
            });
        }
    };


    ////////// Request Threading ////////////////
    private void startRequestThread() {
        synchronized (guard) {
            if (requestThread == null || !requestThread.isRunning()) {
                requestThread = new RequestThread();
                requestThread.start();
            }
            requestThread.notifyRunner();
        }
    }

    private void stopRequestThread() {
        synchronized (guard) {
            if (requestThread != null) {
                requestThread.stop();
            }
        }
    }

    private class RequestThread extends Runner {

        RequestThread() {
            times = Maps.newHashMap();
            delays = Maps.newHashMap();
        }

        @Override
        protected boolean looping() {
            String endpointId = pendingEndpoints.pollFirst();
            long peerId = getPeerId(endpointId);
            Timber.v("Next EndpointId (%s) - PeerId (%d)", endpointId, peerId);
            if (Strings.isNullOrEmpty(endpointId) || peerId == 0L) {
                waitRunner(wait);
                wait += (delayS + delayS);
                return true;
            }
            wait = delayS + delayS;

            //already requested endpoint
            if (states.containsKey(endpointId)) {
                Timber.v("States endpointId (%s) - PeerId (%d) - %s", endpointId, peerId, states.get(endpointId));
            }
            if (states.containsKey(endpointId) && states.get(endpointId) == State.REQUESTING) {
                return true;
            }

            //incoming endpoints: so don't make request on it
            if (directs.containsKey(endpointId) && directs.get(endpointId)) {
                return true;
            }

            if (!times.containsKey(endpointId)) {
                times.put(endpointId, System.currentTimeMillis());
            }

            if (!delays.containsKey(endpointId)) {
                delays.put(endpointId, NearbyUtil.nextRand(5, 15) * delayS);
            }

            if (!requestTries.containsKey(endpointId)) {
                requestTries.put(endpointId, 0);
            }

            boolean readyToRequest = isExpired(times.get(endpointId), delays.get(endpointId));
            Timber.v("Request Attempt %s %s", endpointId, readyToRequest);
            if (readyToRequest) {
                requestConnection(endpointId);
                states.put(endpointId, State.REQUESTING);
                times.remove(endpointId);
                delays.remove(endpointId);
                requestTries.put(endpointId, requestTries.get(endpointId) + 1);
            } else {
                pendingEndpoints.insertLastUniquely(endpointId);
            }
            waitRunner(2 * delayS);
            return true;
        }
    }

    //some public api
    boolean send(long peerId, Payload payload) {
        String acceptedEndpoint = getAcceptedEndpointId(peerId);
        if (acceptedEndpoint == null) {
            return false;
        }
        client.sendPayload(acceptedEndpoint, payload);
        return true;
    }

    private long getPeerId(String endpointId) {
        if (endpoints.inverse().containsKey(endpointId)) {
            return endpoints.inverse().get(endpointId);
        }
        return 0;
    }

    private String getAcceptedEndpointId(long peerId) {
        if (!endpoints.containsKey(peerId)) {
            return null;
        }

        String endpointId = endpoints.get(peerId);
        if (!states.containsKey(endpointId)) {
            return null;
        }
        if (states.get(endpointId) != State.ACCEPTED) {
            return null;
        }
        return endpointId;
    }

    private String toString(long value) {
        if (!cache.containsKey(value)) {
            cache.put(value, String.valueOf(value));
        }
        return cache.get(value);
    }

    private long toLong(String value) {
        if (!cache.containsValue(value)) {
            cache.put(Long.parseLong(value), value);
        }
        return cache.inverse().get(value);
    }
}

File: android/network/src/main/java/com/dreampany/network/nearby/core/Connection.kt
package com.dreampany.network.nearby.core

import android.content.Context
import com.dreampany.network.misc.*
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.nearby.Nearby
import com.google.android.gms.nearby.connection.*
import com.google.common.collect.Maps
import timber.log.Timber
import java.util.concurrent.Executor

/**
 * Created by roman on 7/2/21
 * Copyright (c) 2021 butler. All rights reserved.
 * ifte.net@gmail.com
 * Last modified $file.lastModified
 */
class Connection(
    private val context: Context,
    private val executor: Executor,
    private val callback: Callback
) : ConnectionLifecycleCallback() {

    private enum class State {
        FOUND,
        LOST,
        REQUESTING,
        REQUEST_SUCCESS,
        REQUEST_FAILED,
        INITIATED,
        ALREADY_CONNECTED,
        ACCEPTED,
        REJECTED,
        ERROR,
        DISCONNECTED
    }

    interface Callback {
        fun onConnection(endpointId: String, connected: Boolean)
        fun onPayload(endpointId: String, payload: Payload)
        fun onPayloadStatus(endpointId: String, update: PayloadTransferUpdate)
    }

    private val guard = Object()

    private var client: ConnectionsClient
    private lateinit var advertisingOptions: AdvertisingOptions
    private lateinit var discoveryOptions: DiscoveryOptions

    private var strategy: Strategy? = null
    private var serviceId: String? = null

    @Volatile
    private var advertising = false

    @Volatile
    private var discovering = false

    @Volatile
    private var started = false

    //private val cache: BiMap<Long, String>
    private val states: MutableMap<String, State> // endpointId to State
    private val directs: MutableMap<String, Boolean>  // endpointId to directs (incoming = true or outgoing = false)
    private val requestTries: MutableMap<String, Int>
    private val endpoints: SmartQueue<String>
    private val ENDPOINT_NAME = "ep"

    @Volatile
    private lateinit var runner: com.dreampany.network.misc.Runner

    init {
        client = Nearby.getConnectionsClient(context)

        states = Maps.newConcurrentMap()
        directs = Maps.newConcurrentMap()
        requestTries = Maps.newConcurrentMap()
        endpoints = SmartQueue()
    }

    override fun onConnectionInitiated(endpointId: String, info: ConnectionInfo) {
        if (ENDPOINT_NAME != info.endpointName) return
        //val peerId = info.endpointName
        Timber.v(
            "Connection Initiated[EndpointId:%s][Incoming:%s]",
            endpointId,
            //peerId,
            info.isIncomingConnection
        )
        //endpoints.forcePut(peerId, endpointId)
        states[endpointId] = State.INITIATED
        directs[endpointId] = info.isIncomingConnection
        client.acceptConnection(endpointId, payloadCallback)
    }

    override fun onConnectionResult(endpointId: String, resolution: ConnectionResolution) {
        val accepted = resolution.status.isSuccess

        //val peerId = endpointId.peerId
        Timber.v(
            "Connection Result[EndpointId:%s][Accepted:%s]",
            endpointId,
            //peerId,
            accepted
        )

        states[endpointId] = if (accepted) State.ACCEPTED else State.REJECTED
        endpoints.insertLastUniquely(endpointId)
        executor.execute { startRunner() }

        //if (accepted) {
        //pendingEndpoints.remove(endpointId)
        //if (peerId == null) {
        // TODO
        //} else {
        //    executor.execute { callback.onConnection(peerId, true) }
        //}
        //} else {

        //}
    }

    override fun onDisconnected(endpointId: String) {
        Timber.v("Disconnected [EndpointId:$endpointId]")
        states[endpointId] = State.DISCONNECTED
        endpoints.insertLastUniquely(endpointId)
        //val peerId = endpointId.peerId ?: return
        executor.execute { startRunner() }
    }

    fun start(strategy: Strategy, serviceId: String) {
        synchronized(guard) {
            if (started) return
            this.strategy = strategy
            this.serviceId = serviceId

            started = true

            Timber.v("Starting Nearby Connection %s", strategy.toString())
            advertisingOptions = AdvertisingOptions.Builder().setStrategy(strategy).build()
            discoveryOptions = DiscoveryOptions.Builder().setStrategy(strategy).build()
        }

        startAdvertising()
        startDiscovery()
    }

    fun stop() {
        synchronized(guard) {
            if (started.not()) return
            Timber.v("Stop Connection")
            started = false
            stopRunner()
            stopAdvertising()
            stopDiscovery()
        }
    }

    fun requireRestart(
        strategy: Strategy,
        serviceId: String
    ): Boolean = started
            && (this.strategy != strategy
            || this.serviceId != serviceId)

    fun send(endpointId: String, payload: Payload): Boolean {
        if (!isReady(endpointId)) return false
        client.sendPayload(endpointId, payload)
        return true
    }

    /* private */
    private fun isReady(endpointId: String): Boolean {
        val state = states.get(endpointId) ?: return false
        when (state) {
            State.ALREADY_CONNECTED,
            State.ACCEPTED -> return true
            else -> return false
        }
    }

    /* discovery callback for getting advertised devices */
    private val discoveryCallback = object : EndpointDiscoveryCallback() {
        override fun onEndpointFound(endpointId: String, info: DiscoveredEndpointInfo) {
            val serviceId = info.serviceId
            if (this@Connection.serviceId != serviceId || ENDPOINT_NAME != info.endpointName) {
                Timber.e(
                    "Unknown [ServiceId-EndpointId]:[%s-%s]",
                    serviceId,
                    endpointId
                )
                return
            }

            Timber.v("EndpointFound[EndpointId:%s]", endpointId)

            states[endpointId] = State.FOUND
            requestTries.put(endpointId, 0)
            endpoints.insertLastUniquely(endpointId)
            executor.execute { startRunner() }

            /* priority works: remove old endpoints if exists */
//            if (endpoints.containsKey(peerId)) {
//                val oldEndpointId = endpoints.remove(peerId)
//                states.remove(oldEndpointId)
//                directs.remove(oldEndpointId)
//                pendingEndpoints.remove(oldEndpointId)
//            }
//
//            endpoints.forcePut(peerId, endpointId)
//            states[endpointId] = State.FOUND
//            pendingEndpoints.insertLastUniquely(endpointId)
//            requestTries.put(endpointId, 0)
//            executor.execute { startRequestThread() }
        }

        override fun onEndpointLost(endpointId: String) {
//            val peerId = endpointId.peerId
//            if (peerId == null) {
//                Timber.v("EndpointLost [EndpointId]:[%s]", endpointId)
//                return
//            }

            Timber.v("EndpointLost [EndpointId:$endpointId]")

            states[endpointId] = State.LOST
            directs.remove(endpointId)
            requestTries.remove(endpointId)
            endpoints.remove(endpointId)
            //val peerId = endpointId.peerId ?: return
            //executor.execute {
            //callback.onConnection(peerId, false)
            //}

//            endpoints.remove(peerId)
//            states.put(endpointId, State.LOST)
//            directs.remove(endpointId)
//            pendingEndpoints.remove(endpointId)
//            requestTries.remove(endpointId)
//            executor.execute {
//                callback.onConnection(peerId, false)
//            }
        }
    }

    /* payload callback */
    private val payloadCallback = object : PayloadCallback() {

        override fun onPayloadReceived(endpointId: String, payload: Payload) {
            //val peerId = endpointId.peerId ?: return
            //Timber.v("Payload Received from PeerId (%s)", peerId)
            executor.execute { callback.onPayload(endpointId, payload) }
        }

        override fun onPayloadTransferUpdate(endpointId: String, update: PayloadTransferUpdate) {
            //val peerId = endpointId.peerId ?: return
            //Timber.v("Payload Transfer Update from PeerId (%s)", peerId)
            //executor.execute { callback.onPayloadStatus(peerId, update) }
        }

    }


    private fun startAdvertising() {
        synchronized(guard) {
            if (!started && advertising) return

            Timber.v("Advertising [ServiceId:$serviceId]")
            val serviceId = serviceId ?: return

            client.startAdvertising(ENDPOINT_NAME, serviceId, this, advertisingOptions)
                .addOnFailureListener {
                    advertising = false
                    Timber.e(it, "Failure Advertising [ServiceId:$serviceId]")
                }.addOnSuccessListener {
                    advertising = true
                    Timber.v("Success Advertising [ServiceId:$serviceId]")
                }
        }
    }

    private fun stopAdvertising() {
        synchronized(guard) {
            if (!advertising) return
            advertising = false
            client.stopAdvertising()
        }
    }

    private fun startDiscovery() {
        synchronized(guard) {
            if (!started && discovering) return

            Timber.v("Discovering [ServiceId:$serviceId]")
            val serviceId = serviceId ?: return

            client.startDiscovery(serviceId, discoveryCallback, discoveryOptions)
                .addOnFailureListener {
                    discovering = false
                    Timber.e(it, "Failure Discovering [ServiceId:$serviceId]")
                }.addOnSuccessListener {
                    discovering = true
                    Timber.v("Success Discovering [ServiceId:$serviceId]")
                }
        }
    }

    private fun stopDiscovery() {
        synchronized(guard) {
            if (!discovering) return
            discovering = false
            client.stopDiscovery()
        }
    }

    private fun startRunner() {
        synchronized(guard) {
            if (!::runner.isInitialized) runner = Runner(this)
            if (!runner.isStarted) runner.start()
            runner.notifyRunner()
        }
    }

    private fun stopRunner() {
        synchronized(guard) {
            if (::runner.isInitialized && runner.isStarted) runner.stop()
        }
    }

    private fun requestConnection(endpointId: String) {
        Timber.v("RequestConnection[EndpointId:%s][State:%s]", endpointId, states[endpointId])
        //val peerId = peerId ?: return

        client.requestConnection(ENDPOINT_NAME, endpointId, this)
            .addOnFailureListener {
                Timber.e(it, "Failure RequestConnection[%s]-%s", endpointId, it.message)

                if (it is ApiException) {
                    when (it.statusCode) {
                        ConnectionsStatusCodes.STATUS_ALREADY_CONNECTED_TO_ENDPOINT -> {
                            states[endpointId] = State.ALREADY_CONNECTED
                            endpoints.insertLastUniquely(endpointId)
                            startRunner()
                            return@addOnFailureListener
                        }
                        ConnectionsStatusCodes.STATUS_ENDPOINT_IO_ERROR,
                        ConnectionsStatusCodes.STATUS_RADIO_ERROR,
                        ConnectionsStatusCodes.STATUS_OUT_OF_ORDER_API_CALL -> {
                            states[endpointId] = State.ERROR
                            endpoints.insertLastUniquely(endpointId)
                            startRunner()
                            return@addOnFailureListener
                        }
                    }
                }

                states[endpointId] = State.REQUEST_FAILED
                endpoints.insertLastUniquely(endpointId)
                startRunner()
            }.addOnSuccessListener {
                Timber.v("Success RequestConnection [EndpointId:$endpointId]")
                states[endpointId] = State.REQUEST_SUCCESS
                endpoints.remove(endpointId)
                requestTries.remove(endpointId)
            }
    }

    /* request thread */
    inner class Runner(val connection: Connection) : com.dreampany.network.misc.Runner() {

        @Throws(InterruptedException::class)
        override fun looping(): Boolean {
            val endpointId: String? = connection.endpoints.pollFirst()
            if (endpointId == null) {
                waitRunner(wait)
                wait += delayS
                return true
            }
            wait = delayS

            val state = connection.states.get(endpointId)

            /* already requested endpoint */
            if (state != null) {
                Timber.v("State[EndpointId:$endpointId][State:$state]")
                if (state == State.REQUESTING || state == State.ERROR) return true
                else if (state == State.LOST) {
                    connection.executor.execute {
                        connection.callback.onConnection(
                            endpointId,
                            false
                        )
                    }
                    return true
                } else if (state == State.ACCEPTED || state == State.ALREADY_CONNECTED) {
                    connection.executor.execute {
                        connection.callback.onConnection(
                            endpointId,
                            true
                        )
                    }
                    return true
                }
            }

            /* incoming endpoints: so don't make request on it */
            if (connection.directs.valueOf(endpointId)) return true

            if (!connection.requestTries.containsKey(endpointId)) connection.requestTries.put(
                endpointId,
                0
            )
            if (!times.containsKey(endpointId)) times[endpointId] = currentMillis
            if (!delays.containsKey(endpointId)) delays[endpointId] = Utils.nextRand(2, 5) * delayS

            if (times.timeOf(endpointId).isExpired(delays.valueOf(endpointId))) {
                connection.requestConnection(endpointId)
                connection.states[endpointId] = State.REQUESTING
                times.remove(endpointId)
                delays.remove(endpointId)
                connection.requestTries.let {
                    it.put(endpointId, it.valueOf(endpointId).inc())
                }
            }
            connection.endpoints.insertLastUniquely(endpointId)
            waitRunner(delayS)
            return true
        }

    }

}